src/agreement.rs

```diff
@@ -1,42 +1,283 @@
+                outgoing.push_back(output_message);
+            }
+            Ok((coin_result.0, outgoing))
+        } else {
+            Ok((None, outgoing))
+        }
+    }
+
+    /// AUX_r messages such that the set of values carried by those messages is
+    /// a subset of bin_values_r. Outputs this subset.
+    ///
+    /// FIXME: Clarify whether the values of AUX messages should be the same or
+    /// not. It is assumed in `count_aux` that they can differ.
+    ///
+    /// In general, we can't expect every good node to send the same AUX value,
+    /// so waiting for N - f agreeing messages would not always terminate. We
+    /// can, however, expect every good node to send an AUX value that will
+    /// eventually end up in our bin_values.
+    fn count_aux(&self) -> (usize, BTreeSet<bool>) {
+        let vals = BTreeSet::new();
+        (
+            self.received_aux
+                .values()
+                .filter(|values| values.is_subset(&self.bin_values))
+                .map(|values| vals.union(values))
+                .count(),
+            vals,
+        )
+    }
+
+    /// Waits until at least (N − f) AUX_r messages have been received, such that
+    /// the set of values carried by these messages, vals, are a subset of
+    /// bin_values_r (note that bin_values_r may continue to change as BVAL_r
+    /// messages are received, thus this condition may be triggered upon arrival
+    /// of either an AUX_r or a BVAL_r message).
+    ///
+    /// `try_coin` outputs an optional combination of the agreement value and
+    /// the agreement broadcast message.
+    fn try_coin(&mut self) -> (Option<bool>, Option<AgreementMessage>) {
+        let (count_aux, vals) = self.count_aux();
+        if count_aux < self.num_nodes - self.num_faulty_nodes {
+            // Continue waiting for the (N - f) AUX messages.
+            (None, None)
+        } else {
+            // FIXME: Implement the Common Coin algorithm. At the moment the
+            // coin value is random and local to each instance of Agreement.
+            let coin2 = random::<bool>();
+
+            // Check the termination condition: "continue looping until both a
+            // value b is output in some round r, and the value Coin_r' = b for
+            // some round r' > r."
+            self.terminated = self.terminated || self.estimated.values().any(|b| *b == coin2);
+
+            // Prepare to start the next epoch.
+            self.bin_values.clear();
+
+            if vals.len() != 1 {
+                // Start the next epoch.
+                self.epoch += 1;
+                (None, Some(self.set_input(coin2)))
+            } else {
+                let mut message = None;
+                // NOTE: `vals` has exactly one element due to `vals.len() == 1`
+                let output: Vec<Option<bool>> = vals.into_iter()
+                    .take(1)
+                    .map(|b| {
+                        message = Some(self.set_input(b));
```

Comment:
Does `self.epoch` need to be incremented before this call? I imagine the `BVal` messages this sends should have the _next_ epoch.

Reply:
You are right. It makes sense to increment the epoch earlier.